### ISSUE TITLE
Frontend component for displaying a schedule to viewers

### ISSUE THREAD

**Issue by gabek** [owncast/owncast#2372]
### Share your bug report, feature request, or comment.

Weekly view, monthly view, list view.

**Comment by germainelee** [comment]
For the initial release, do you want the admin to have a toggle to display stream schedule to viewers? Or the schedule tab is shown always even when there aren't any items in the schedule? The latter might cause viewers be disappointed when they constantly see an empty schedule, when the admin actually doesn't plan on using the scheduling feature any time soon.

**Comment by gabek** [comment]
Yeah, it's a feature that should need to be turned on. There are many types of Owncast instances that would never have a schedule (24/7 streams), so showing one wouldn't make any sense.

**Comment by germainelee** [comment]
I disagree that viewers should be offered the option to switch multiple views to see the schedule despite because it is common that streamers have inconsistent schedule (e.g. don't stream every week and consistently due to personal reasons). So it might not leave a good impression to audience who stumbled across the instance via the Directory. A list / card view of upcoming events is sufficient to inform users that there events planned and when they can come back to watch.

**Comment by gabek** [comment]
If a streamer turns on scheduled events but never has any events scheduled, leading to an empty calendar, that's their fault. They should turn off the feature if that's a better experience for their use case. We shouldn't be dumbing the functionality down in case people don't use it correctly.

An agenda/list view is great for seeing a list of events all at once, but if you want a monthly overview, for example, a calendar view of the month is the best way to view that. A list won't make it obvious that something is every Tuesday, but a calendar view will. People are used to being able to look at schedules in different ways.

This is the component I'm thinking should be used: https://www.npmjs.com/package/react-big-calendar

<blockquote><img src="https://static-production.npmjs.com/338e4905a2684ca96e08c7780fc68412.png" width="48" align="right"><div><img src="https://static-production.npmjs.com/da3ab40fb0861d15c83854c29f5f2962.png" height="14"> npm</div><div><strong><a href="https://www.npmjs.com/package/react-big-calendar">react-big-calendar</a></strong></div><div>Calendar! with events. Latest version: 1.13.1, last published: 3 hours ago. Start using react-big-calendar in your project by running `npm i react-big-calendar`. There are 284 other projects in the npm registry using react-big-calendar.</div></blockquote>

**Comment by germainelee** [comment]
There should not be a correct way to use the feature for different use case. A simple message informing viewers that he does not have any streams planned at the moment is sufficient. As it immediately informs the viewer that there isn't any event scheduled compared to the calendar where viewer is searching through to see if there is any scheduled events planned in the future.

A card view is also great for seeing upcoming events as card can provide a snapshot of an event, making it easy for users to quickly understand what the event is about. The event description comes in handy here. The events in the calendar has a start and end time. But some streamers won't have a definite end time because they are not sure how long they want to stream each time.

If the user sees card A is Tue, Jul 16 @ 2pm , card B is Tue, Jul 23 @ 2pm , card C is Tue, Jul 30 @ 2pm . He can make a smart guess that the streamer streams every Tue. Therefore, Cards / list view is sufficient in succinctly communicating upcoming events

**Comment by gabek** [comment]
Events are expected to have a start and end time, that's what a schedule is.

The schedule components should not disappear from the page unless they're turned off, and it's up to the admin to turn them off. We shouldn't be automatically turning them off just because the schedule is empty. That's up to them to decide. It would feel broken if things just appear and disappear from the page, there should be consistency if the feature is enabled.

There's no downside to letting people view the schedule as a calendar if they'd like to, that is a natural way to look at a schedule and is what people are used to. Of course, there would be the agenda view as well, because that is also something that people are used to and would expect. Those two views are used for different reasons, and both are valid. This is a schedule and should be able to be browsed like a schedule.

**Comment by germainelee** [comment]
> Events are expected to have a start and end time, that's what a schedule is.

Events such as conferences and church mass usually have a start and end time, but gaming streams can be unpredictable, and setting a strict end time might limit spontaneity and the ability to extend the stream. So end time should be optional rather than mandating it. This accommodates different styles of streaming.

Although it is common to view schedules in a calendar format, it might not have the intended user behaviour here. The viewer should be able to quickly glance at the upcoming events below to see all necessary details like date, time, and event description and can quickly find the events that are of interest without having to navigate through multiple pages.

Therefore the card view ensures that viewers are well-informed about upcoming event as it is more condensed and in a easy-to-read format.

**Comment by gabek** [comment]
I was just using a piece of software that uses a calendar that works pretty much exactly like ours would. In this case, it's July 19th and there are no future scheduled items. But one thing I never thought of until I was looking at this was that even though nothing is scheduled into the future, I still might want to look into the past. Seeing the different events that happened in the past could be interesting to see on the calendar. 

![image](https://github.com/user-attachments/assets/35db61e7-881b-4845-9f1f-e709255a8de1)

**Comment by germainelee** [comment]
imo week view is sufficient. This is what I have in mind. It displays the events similar to the admin view. If that week is empty, user will see a "No upcoming streams at this time." message.
![schedule FE](https://github.com/user-attachments/assets/4d561d3f-4c90-49e7-8101-7ddcc9304247)

**Comment by germainelee** [comment]
How would the calendar look in mobile view? Can you share a screenshot as I can't seem to find it in the [react-big-calendar](https://github.com/jquense/react-big-calendar?tab=readme-ov-file#readme) docs

**Comment by gabek** [comment]
That's a really good question. Maybe we could be simple and only do an agenda/list view for mobile.

**Comment by germainelee** [comment]
That's a good option. We should take into consideration of the resizable chat panel plus small desktop width. So in order to make it responsive and having no awkward transitions, I'm thinking to either do
(A) compressed calendar, clicking on the date opens a details popup 
![Image](https://github.com/user-attachments/assets/e3bbf97a-762b-46b6-8f56-ba25e86d415e)

(B) Mobile-first calendar, with calendar on top and details card below/beside depending on the breakpoint
![Image](https://github.com/user-attachments/assets/a85faca2-c712-4085-9c95-2f2a9cc9fe70)

**Comment by gabek** [comment]
The problem with building a brand new custom calendar, is somebody has to build a brand new custom calendar from scratch. That's not a small thing. That's why I've pushed really hard to not have to do it. That looks great, but I'm pretty sure it would never get done. A calendar component is a project in itself.

**Comment by germainelee** [comment]
I see... so if we were to use _react-big-calendar_, the only customization would be the color ig?

**Comment by gabek** [comment]
We'd have quite a bit of frontend flexibility using CSS with an existing calendar to change how it looked. But not so much how it worked.

**Comment by germainelee** [comment]
How about standardizing the schedule to a weekly agenda/list view for both mobile and desktop? 
![Image](https://github.com/user-attachments/assets/dcf392de-f339-4715-9bd7-6f4a4881428d)

**Comment by gabek** [comment]
I think for mobile having a single week view in a list is fine (it also doesn't need to be windowed to just a week, it can be the agenda view). But for desktop a full month view is nice, I don't think it should be removed just because mobile won't have it.
